### ISSUE TITLE
Episodic batch-wide degenerate output (endless BOS token spam) with DSpark spec-decode on DeepSeek-V4-Flash-0731; draft acceptance collapses to exactly 0.0%

### ISSUE THREAD

**Issue by repandv** [voipmonitor/rtx6kpro#53]
**Environment**

Image: voipmonitor/vllm:fathomless-firmament-v16-vllm8f86f42-b12xfe06f49-fi801d57a-cu132-20260714 (v0.11.2.dev280)
Model: DeepSeek-V4-Flash-0731 official release, native MXFP4, served via MODE=dspark (built-in DSpark module, same checkpoint as draft source)
HW: 4× RTX PRO 6000 Blackwell SE 96GB (SM120a), TP=4, PCIe (PEX89144 switches)
Config: --kv-cache-dtype fp8 (fp8_ds_mla), --enable-prefix-caching, --max-num-seqs 64, --max-num-batched-tokens 8192, --max-cudagraph-capture-size 384, --attention-backend B12X_MLA_SPARSE, spec-config {"method":"dspark","num_speculative_tokens":5,"draft_sample_method":"probabilistic","rejection_sample_method":"standard"}, thinking=true, reasoning_effort=high
Context: 262144; mixed agentic + chat production traffic, ~10-30 concurrent streams peak

**Symptom**

Twice within the first ~26h of production traffic on 0731 (never seen on the DSpark preview checkpoint with the identical config over 2 weeks), the engine entered an episodic failure mode:

Multiple concurrent streams (different users, different clients, unrelated prompts) simultaneously began emitting endless <｜begin▁of▁sentence｜> repetitions. Affected responses were garbage from the very first generated token (sample below — note it lands inside the reasoning block). One stream produced 155,033 completion tokens of pure BOS at ~330 tok/s before finishing.
During the episode, SpecDecoding metrics show a distinctive signature — acceptance collapses to exactly 0.000 across all 5 positions and stays there for the entire episode (~20 min), with drafted throughput ~3300 tok/s and accepted = 0:
18:29:00 ... Per-position acceptance rate: 0.958, 0.861, 0.777, 0.723, 0.639, Avg: 79.2%   <- anomalous spike (looping stream: trivially predictable)
18:29:30 ... Per-position acceptance rate: 0.305, 0.217, 0.160, 0.115, 0.077, Avg: 17.5%
18:29:40 ... Per-position acceptance rate: 0.000, 0.000, 0.000, 0.000, 0.000, Avg: 0.0%    <- stays here ~20 min

Sequence: normal (40-58%) → brief spike to 79% (a stream starts looping; drafter predicts the loop perfectly) → collapse to strict 0.0% batch-wide. Healthy streams sharing the batch were starved (~25 tok/s each). The episode self-resolved when the runaway streams drained (client aborts + max-token stops). Episode 2 (next day, ~7-8% acceptance sustained) required a container restart. No errors, no tracebacks, no Xid/NCCL/watchdog events in either episode; the engine believed itself healthy throughout.

Trigger correlation: episode 1 started seconds after a burst admission (Running 1→25, prefill spike ~9900 tok/s) and JIT kernel compilations were observed mid-episode (UnifiedDecodeKernel call_pertok/call_extra_pertok, extra_topk_bucket=2048, chunks_per_split=34, disk-cache-miss — first execution of these deep-context decode shapes after 44h uptime). Long-context requests (190-230K prompts) were present in the batch.

**Sample output** (Open WebUI chat DB, verbatim; response began like this from token 0 and continued for 5.3 MB):

<details type="reasoning" done="true" duration="463">
<summary>Thought for 463 seconds</summary>
> <｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>...

**Two hypotheses** (we could not discriminate from logs alone):

A. _Proposer-side_: in worker/gpu/spec_decode/dspark/speculator.py, invalid draft samples are replaced via torch.zeros_like — i.e. with token id 0, which is BOS for this tokenizer. Normally these placeholders are trimmed by valid_lengths/capacity before verification; if any inconsistency in the ragged-batch bookkeeping lets them (or mixed-context anchors — the anchor read input_ids[::num_query_per_req] assumes a strictly uniform query-block layout) reach acceptance, BOS floods the output. The 0.0%-flat acceptance during the episode shows the drafter and target fully decorrelated, consistent with per-request state corruption under concurrency.

B. _Target-side_: uncaptured/unwarmed deep-context sparse-MLA decode paths (the JIT'd *_pertok shapes) producing corrupted logits under a concurrent ragged batch; BOS-argmax garbage then explains both the output and the 0% acceptance (drafter can't predict junk). We have previously seen a paged sparse-MLA concurrency>1 corruption on different hardware (SM80) with this model family, so this is not hypothetical for us.

**Current status / workaround**

Running the same checkpoint with MODE=mtp0 (spec decode disabled entirely) since the second episode; no recurrence so far (will update after 24-48h soak). We also set a default max_new_tokens cap as damage control. Happy to provide full logs, the Bifrost gateway per-request statistics for both episodes, or run instrumented builds — we can reproduce production-like concurrency on this host.

**Comment by voipmonitor** [comment]
## r27 status: release validation passed, production soak still required

r27 keeps fixed K5 as the DS4 default and passed the following exact-image checks:

- 100/100 numerically checked outputs across 50 repeated CC2 batches;
- 6/6 independent concurrent approximately 50k-token prefills;
- a clean uncached 65,539-token prefill;
- post-run health and error-signature scan.

Image:

```text
voipmonitor/vllm:gilded-gnosis-v20-vllm966d57c-sibbbdccc-fi801d57a-cu132-20260803-r27
manifest: sha256:2605fda01797f33239af4c95ec7449505fe57d9b9de9687792f5b8273d3201a7
```

Runbook:
https://github.com/local-inference-lab/rtx6kpro/blob/master/models/ds4dspark-v20.md

This does **not** close the issue. The reported failure is episodic under 10-30 concurrent production streams, approximately 190-230k prompts, and a long soak; the release gate is much shorter. vLLM #235 repairs the 0731 prompt/tool contract but is not presented as a fix for this batch-wide runtime state failure. Please use this image for the next instrumented soak and retain acceptance telemetry plus the full launch configuration if the strict 0.0% episode recurs.

**Comment by wwwcxc5120** [comment]
DSpark + thinking 高并发 BOS 重复
问题： 0731 权重 + DSpark 投机解码 + thinking 模式在高并发下，reasoning 字段出现大量重复 <｜begin▁of▁sentence｜> BOS token（62-146 个连续），content 为空，max_tokens 停止条件失效。概率性 10-75%，40 并发 + max_tokens=150 时 65%。

根因（2026-08-01 实证）： DSpark 批处理失同步 bug——thinking 块结束→content 转移点的批处理失同步，draft 生成但验证系统性拒绝（接受率 42%→0.8% 且不可逆）。官方文档记录 "if thinking is false, even with mtp there is no problem"。

关键结论：

唯一 0% 病态配置：reasoning_effort=none（关闭 thinking）——但用户要求保留 thinking
参数无法根除 ：capacity 削减（cap1，三轮 10/5/60%）、dynamic 深度（cap2，三轮 0/60/30%）均无效
阶段一方案（2026-08-01，已退位）：MODE: dspark-mtp0（0731 权重 + 关闭 DSpark 投机解码 = 纯自回归）
speculative_config=None → 批处理失同步根源消失 → thinking 完整保留
验证：40 并发×3 轮 0% / 80 并发×40 请求×2 轮 0% / Anthropic 路径 0%
吞吐损失 ~20-30%（峰值 364 tok/s，因 DSpark 接受率本就仅 8-12%，投机收益有限）
根因认知修正：BOS 病态根源是 DSpark draft 头结构（Eagle3）而非投机解码本身——换用标准 MTP draft 头后投机 + thinking 共存无病态
验证：3 轮×64 并发 0% 病态 / thinking 100% 保留 / 192/192 成功 / 错误日志 0
重新获得投机解码加速（短请求均值 176.4 tok/s）且无病态

**Comment by hassellof** [comment]
Your `num_speculative_tokens: 5` is, on our hardware, the one draft depth that is broken — and depths either side of it are clean.

We run the same GPUs (4× RTX PRO 6000 Blackwell, sm_120) but on SGLang rather than vLLM, TP=4 + DP attention + EP=4, 0731 weights, thinking on, streaming, agentic tool traffic. We saw the same class of failure — degenerate output mid-generation, no errors, engine believes it is healthy — and swept the DSpark draft depth under an identical 4-phase bursty load instead of testing one value:

| draft depth | requests | corruption events |
|---|---|---|
| 3 | 10,885 (5 cold launches) | 0 |
| 4 | 11,040 (5 cold launches) | 0 |
| **5** | 2,148 / 2,138 | **10 / 7** |
| 6 | 2,282 | 0 |
| 7 | 2,198 | 0 |

Depth **5** is the only broken value. It is not monotonic — 6 and 7 are clean — so it is not a capacity, window or ring-size bound. What is special about 5 is that it is exactly `dspark_block_size` in the 0731 checkpoint config: the draft head's *native* block size, and the depth you land on whenever the stack infers it rather than being told. Notably this also matches your "never seen on the DSpark preview checkpoint with the identical config" — the preview ships a different draft module.

Your per-position table is 5 wide, so you are at depth 5. **The cheap experiment is `num_speculative_tokens: 7`, not spec-decode off.** If 7 is clean for you too, you keep speculative decoding and you have narrowed this to the native-block-size path — which would also be a strong signal that the bug is in the shared DSpark semantics rather than in either engine's plumbing.

On your two hypotheses, from what we could measure on our side:

- **A (proposer-side)** matches our evidence better. We instrumented SGLang's plan kernel and confirmed the verify window does reach it, and we still get corruption at depth 5 only. The `torch.zeros_like` → token id 0 → BOS chain you describe is a very plausible mechanism for the specific *shape* of your garbage, and BOS spam is a phenotype we did not see (ours is repetition collapse, token salad and `</think>` leaking into content) — which may just be a difference in how each engine fills invalid draft slots.
- **B (target-side deep-context decode)** we could not confirm. Our corruption reproduces at ordinary context lengths, well below the 190-230K prompts in your batch, and does not need a JIT compile to be in flight. It also disappears entirely at depth 6 with the same kernels, which is hard to square with an uncaptured decode shape.

The 79% acceptance spike immediately before the collapse is a detail worth keeping: we saw the same ordering — the drafter locks onto a stream that has already started repeating (trivially predictable, acceptance spikes), and the collapse follows. That is consistent with the corruption *causing* the acceptance anomaly rather than the reverse, which matters for anyone tempted to treat the acceptance metric as the alarm.

Things we falsified by measurement, so nobody repeats them: c4 ring capacity, verify window > `compress_ratio`, the `kMaxMTPDraftTokens = 4` row budget, and NaN logits. Full evidence and repro protocol: https://github.com/sgl-project/sglang/issues/33800

Our SM120 recipe, patch set and measured numbers (SGLang side, but the depth finding should transfer): https://github.com/ombori/deepseek-v4-flash-0731-sglang-4x-rtx-pro-6000

Happy to run candidate patches on our box — each arm of the repro takes about 30 minutes.

**Comment by voipmonitor** [comment]
@hassellof your bot did wrong assumption to report sglang problem - we are using vllm 

@repandv report issues for the latest docker images not the old one

**Comment by hassellof** [comment]
@voipmonitor Apologies if the framing was unclear, no assumption about your stack intended. We know #53 is vLLM; our numbers are from SGLang on the same GPUs, offered precisely because the depth-5 signature showing up under a second engine is what makes it look like a property of the DSpark draft head rather than of either engine's plumbing. The actionable part for this issue is unchanged and engine-agnostic: repandv's config runs num_speculative_tokens: 5, and the cheap discriminating experiment is 7. If 7 is clean under vLLM too, that's the cross-engine confirmation, and it keeps spec decode instead of giving it up. But I see you closed the issue so maybe already solved.
